src/spicy/runtime-support.h

```diff
@@ -929,36 +920,35 @@ inline ValPtr to_val(const T& t, TypePtr target) {
     return rval;
 }
 
-/**
- * Converts a Spicy-side enum to a Zeek record value. The result is returned
- * with ref count +1.
- */
-template<typename T, typename std::enable_if_t<std::is_enum<typename T::Value>::value>*>
-inline ValPtr to_val(const T& t, TypePtr target) {
-#if SPICY_VERSION_NUMBER >= 10700
-    auto proto = typename T::Value(t.value());
-#else
-    auto proto = t;
-#endif
-
-    return to_val(proto, target);
+/** Maps HILTI's `Protocol` enum to Zeek's `transport_proto` enum. */
+inline ValPtr to_val_for_transport_proto(int64_t val, const TypePtr& target) {
+    switch ( val ) {
+        case hilti::rt::Protocol::TCP: return id::transport_proto->GetEnumVal(::TransportProto::TRANSPORT_TCP);
+        case hilti::rt::Protocol::UDP: return id::transport_proto->GetEnumVal(::TransportProto::TRANSPORT_UDP);
+        case hilti::rt::Protocol::ICMP: return id::transport_proto->GetEnumVal(::TransportProto::TRANSPORT_ICMP);
+        default: return id::transport_proto->GetEnumVal(::TransportProto::TRANSPORT_UNKNOWN);
+    }
+
+    hilti::rt::cannot_be_reached();
 }
 
 /**
- * Converts a C++ Spicy-side enum to a Zeek record value. The result is returned
- * with ref count +1. This specialization is provided for compatibility with <spicy-1.7.0.
- *
- * TODO(bbannier): remove this once we drop support for Spicy versions before 1.7.0.
+ * Converts a Spicy-side enum to a Zeek enum value. The result is returned
+ * with ref count +1.
  */
-template<typename T, typename std::enable_if_t<std::is_enum<T>::value>*>
-inline ValPtr to_val(const T& t, TypePtr target) {
+template<typename T, typename std::enable_if_t<std::is_enum<typename T::Value>::value>*>
+inline ValPtr to_val(const T& t, const TypePtr& target) {
     if ( target->Tag() != TYPE_ENUM )
         throw TypeMismatch("enum", target);
 
     // We'll usually be getting an int64_t for T, but allow other signed ints
     // as well.
-    static_assert(std::is_signed<std::underlying_type_t<T>>{});
-    auto it = static_cast<int64_t>(t);
+    static_assert(std::is_signed<std::underlying_type_t<typename T::Value>>{});
+    auto it = static_cast<int64_t>(t.value());
+
+    if ( target->GetName() == "transport_proto" )
+        // Special case: map enum values to Zeek's semantics.
+        return to_val_for_transport_proto(it, target);
```

Comment:
We can now convert any Spicy enum to a `transport_proto`. That seems at the very least fragile.

I'd suggest that we 1) check that we actually have a port value (either by adding an additional condition to this conditional, or by throwing an error if we see a mismatching type), and pass a proper `Protocol::Value` to `to_val_for_transport_proto` instead of an `int64_t`, e.g., here

```suggestion
    // Special case: map enum values to Zeek's semantics.
    if ( target->GetName() == "transport_proto" ) {
        if ( ! std::is_same_v<T, hilti::rt::Protocol::Value> )
            throw TypeMismatch(hilti::rt::demangle(typeid(t).name()), target);
        return to_val_for_transport_proto(static_cast<hilti::rt::Protocol::Value>(it), target);
    }
```